pynotify/models.py

```diff
@@ -141,19 +149,50 @@ class Meta:
         verbose_name_plural = _l('notifications')
         ordering = ('-created_at',)
 
+    def _render(self, field):
+        return self.template.render(field, self.context)
+
+    def _pre_save(self, *args, **kwargs):
+        keys = set(self.get_extra_data()) & set(obj.name for obj in self.related_objects.all())
+        if keys:
+            raise ValueError('Related objects and extra data contain same key(s): {}'.format(', '.join(keys)))
+
     @cached_property
     def related_objects_dict(self):
         """
         Returns related objects as a dictionary where key is name of the related object and value is the object itself.
-        This dictionary is used for rendering template fields.
         """
         output = {}
         for obj in self.related_objects.all():
             output[obj.name] = obj.content_object
         return output
 
-    def _render(self, field):
-        return self.template.render(field, self.related_objects_dict)
+    @property
+    def context(self):
+        """
+        Returns context dictionary used for rendering the template.
+        """
+        return {**self.related_objects_dict, **self.get_extra_data()}
+
+    def set_extra_data(self, extra_data):
+        """
+        Setter for ``extra_data`` field.
+
+        Arguments:
+            extra_data: Dictionary of JSON serializable values.
+        """
+        if not isinstance(extra_data, dict):
+            raise ValueError('Extra data must be a dictionary.')
+        self.extra_data = json.dumps(extra_data)
```

Comment:
You don't want to use JsonField? Reasons? Are you scary of PostgreSQL DB restrictions?

Reply:
- The main reason is I don't want to force users to use PostgreSQL
- Example project included in this library is built on SQLite, so it can be quickly tried without installing and configuring a full blown DB system like PostgreSQL.
- The same way of storing extra data is also used in verification token library here in druids, and it has proven to work well in production environment
- The performace benefit of JsonField is imho negligible here

Reply:
OK It makes sense. Please merge it